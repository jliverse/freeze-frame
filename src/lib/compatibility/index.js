// @license http://opensource.org/licenses/MIT
// copyright Paul Irish 2015
(function(e) {
  if ('performance' in e === false) {
    e.performance = {};
  }
  if ('now' in e.performance === false) {
    const { timing } = e.performance;
    let nowOffset = (timing && timing.navigationStart) || Date.now();
    e.performance.now = () => Date.now() - nowOffset;
  }
})(window);
