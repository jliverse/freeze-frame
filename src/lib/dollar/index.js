class Dollar {
  constructor(el = document.body) {
    this.el = el && el.nodeType ? el : document.querySelector(el) || {};
    this.classNames = (this.el.className && this.el.className.split(' ')) || [];
  }

  hasClass(className) {
    return this.classNames.indexOf(className) >= 0;
  }

  addClass(className) {
    if (this.classNames.indexOf(className) < 0) {
      this.classNames.push(className);
      this.el.className = this.classNames.join(' ');
    }
    return this;
  }

  attr(name, value) {
    if (value) {
      this.el.setAttribute(name, value);
      return this;
    } else {
      return this.el.getAttribute(name);
    }
  }

  css(name) {
    if (!this.el.nodeType) {
      return undefined;
    }
    const style = window.getComputedStyle(this.el);
    if (!style) {
      return undefined;
    }
    switch (name) {
      case 'color':
        return colorAsHex(style.getPropertyValue('color'));
      default:
        return style.getPropertyValue(name);
    }
  }

  removeClass(className) {
    const index = this.classNames.indexOf(className);
    if (index >= 0) {
      this.classNames.splice(index, 1);
      this.el.className = this.classNames.join(' ');
    }
    return this;
  }

  toggleClass(className) {
    return hasClass(className) ? removeClass(className) : addClass(className);
  }
}

export default el => {
  return new Dollar(el);
};
