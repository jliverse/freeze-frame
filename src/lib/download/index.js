import FileSaver from 'file-saver';

export function download(
  data,
  filename,
  mimeType = 'application/octet-stream'
) {
  FileSaver.saveAs(new Blob([data], { type: mimeType }), filename);
}
