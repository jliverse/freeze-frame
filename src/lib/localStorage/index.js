const ls = window.localStorage;

const ERROR_LOCAL_STORAGE_NOT_AVAILABLE =
  'The localStorage API is not available on this device.';
export class LocalStorage {
  isAvailable() {
    if (!ls) {
      return false;
    }
    const _ = '_';
    try {
      ls.setItem(_, _);
      return true;
    } catch (e) {
      return false;
    } finally {
      ls.removeItem(_);
    }
  }

  sizeInBytes() {
    let size = 0;
    for (let key in ls) {
      switch (key) {
        case 'clear':
        case 'getItem':
        case 'key':
        case 'length':
        case 'removeItem':
        case 'setItem':
          continue;
      }
      size += ls[key].length;
    }
    return size;
  }

  withKey(key) {
    if (!this.isAvailable()) {
      console &&
        console.warn &&
        console.warn(ERROR_LOCAL_STORAGE_NOT_AVAILABLE);
      throw new Error(ERROR_LOCAL_STORAGE_NOT_AVAILABLE);
    }
    return {
      get: () => JSON.parse(ls.getItem(key)),
      set: value => ls.setItem(key, JSON.stringify(value)),
      remove: () => ls.removeItem(key),
      key: () => key,
      getAndRemove: () => {
        try {
          return JSON.parse(ls.getItem(key));
        } finally {
          ls.removeItem(key);
        }
      }
    };
  }
}

const instance = new LocalStorage();
module.exports = instance;
module.exports.isAvailable = instance.isAvailable.bind(instance);
module.exports.sizeInBytes = instance.sizeInBytes.bind(instance);
module.exports.withKey = instance.withKey.bind(instance);
