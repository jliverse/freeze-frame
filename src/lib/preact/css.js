export function css() {
  return [...arguments].filter(Boolean).join(' ');
}
