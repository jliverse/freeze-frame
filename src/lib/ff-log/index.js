import 'lib/compatibility';
import isObject from 'lodash/isObject';

const DEFAULT_TIMER = window.performance;

export default class Log {
  data = [];

  constructor(timer = DEFAULT_TIMER) {
    this.timer = timer;
    this.start = timer.now();
    this.now = () => Math.round(this.timer.now() - this.start);
  }

  log(data) {
    const time = this.now();
    if ('debug' in console) {
      console.debug(`‚è≤ ${time} ms.`, [...arguments]);
    }
    if (isObject(data)) {
      this.data.push({ time, ...data });
    } else {
      this.data.push({ time, text: data });
    }
  }
}
