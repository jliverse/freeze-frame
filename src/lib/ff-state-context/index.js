import 'lib/compatibility';

import cloneDeep from 'lodash/cloneDeep';
import fromPairs from 'lodash/fromPairs';
import isFunction from 'lodash/isFunction';
import once from 'lodash/once';
import partial from 'lodash/partial';
import toPairs from 'lodash/toPairs';

import { cycle, randomBinomials } from '../random';
import { uuid } from '../uuid';

const ANIMATION_CYCLE_DELAY = 2000;
const MINIMUM_BEFORE_CUE_IN_MILLIS = 1000;
const MAXIMUM_BEFORE_CUE_IN_MILLIS = 2000;
const DELAY_AFTER_CUE = 600;
const AWAY_DURATION = 1000;
const REWARD_DURATION = 2400;
const PUNISH_DURATION = 0;

class State {
  __state__ = true;
  uuid = uuid();
}

export default class {
  delegate = {};
  state = {};
  options = {
    lookCriterion: 2,
    trialCriterion: 80,
    cueDurationInitial: 200,
    cueDurationIncrement: 40,
    cueDurationMaximum: 1200
  };

  directions = cycle(randomBinomials, 4);

  constructor(options = {}) {
    this.setOptions(options);
  }

  getState = () => cloneDeep(this.state);

  start = () => {
    if (this.isStarted()) {
      return;
    }
    this.hasFocus = false;
    this.state = {
      isStarted: true,
      trialNumber: 0,
      cueDuration: this.options.cueDurationInitial,
      isCalibrating: true,
      isAnimating: true,
      isDistracting: false,
      lookNumber: 0
    };
    const {
      didExperimentStart,
      didAnimationStart,
      didStateChange
    } = this.delegate;
    isFunction(didExperimentStart) && didExperimentStart(this.getState());
    this.animationTimer = this.createAnimationTimer();
    isFunction(didAnimationStart) && didAnimationStart({ ...this.getState() });
    isFunction(didStateChange) && didStateChange(this.getState());
  };

  stop = () => {
    const { didExperimentEnd } = this.delegate;
    const { isStarted } = this.state;
    if (isStarted) {
      this.state = { ...this.state, isStarted: false };
      isFunction(didExperimentEnd) && didExperimentEnd(this.getState());
      clearInterval(this.animationTimer);
    }
  };

  keyDown = () => {
    if (!this.isStarted()) {
      return;
    }
    if (this.state.trialNumber === 0) {
      this.state.trialNumber++;
      this.state.cueDirection = this.nextDirectionAsBoolean();
      const { didStateChange } = this.delegate;
      isFunction(didStateChange) && didStateChange(this.getState());
    }
    if (!this.hasFocus) {
      this.nextTrial();
      this.hasFocus = true;
    }
  };

  keyUp = () => {
    this.hasFocus = false;
    isFunction(this.keyDelegate) && this.keyDelegate(false);
  };

  blur = () => (this.hasFocus = false);

  createAnimationTimer = () => {
    return setInterval(() => {
      if (!this.isInterestingTrial()) {
        return;
      }
      const { didAnimationChange } = this.delegate;
      isFunction(didAnimationChange) && didAnimationChange(this.getState());
    }, 2000);
  };

  isStarted = () => !!this.state.isStarted;

  isAnimating = () => this.state.isAnimating;

  isDistracting = () => this.state.isDistracting;

  getSaccadeAwayThreshold = () => this.state.cueDuration + AWAY_DURATION;

  getSaccadeThreshold = () => DELAY_AFTER_CUE;

  getDelayInMilliseconds = () =>
    MINIMUM_BEFORE_CUE_IN_MILLIS +
    (MAXIMUM_BEFORE_CUE_IN_MILLIS - MINIMUM_BEFORE_CUE_IN_MILLIS) *
      Math.random();

  nextDirectionAsBoolean = () => ::this.directions.next().value;

  isBoringTrial = () => this.state.trialNumber % 2 === 1;

  isInterestingTrial = () => (this.state.trialNumber || 0) % 2 === 0;

  reset() {
    clearInterval(this.animationTimer);
    this.state = {};
    this.nextTrial = once(::this._nextTrial);
    const { didStateChange } = this.delegate;
    isFunction(didStateChange) && didStateChange(this.getState());
  }

  setOptions = options => {
    const numericOptions = fromPairs(toPairs(options).map(([i, j]) => [i, +j]));
    this.options = { ...this.options, ...numericOptions };
    this.reset();
  };

  setCueDuration = cueDuration => {
    this._calibrateWithDuration(cueDuration);
  };

  _calibrateWithDuration = (cueDuration, options) => {
    if (!this.state.isCalibrating) {
      return;
    }
    this.state.cueDuration = cueDuration;
    this.state.isCalibrating = false;
    this.state.trialNumberAtCalibrationEnd =
      Math.max(0, this.state.trialNumber) + this.options.trialCriterion;
    const { didCalibrationEnd } = this.delegate;
    isFunction(didCalibrationEnd) &&
      didCalibrationEnd(this.getState(), options);
  };

  _incrementCueDuration = () => {
    // Reward.
    if (!this.state.isCalibrating) {
      return;
    }
    this.state.cueDuration += this.options.cueDurationIncrement;
    this.state.lookNumber = 0;

    if (this.state.cueDuration > this.options.cueDurationMaximum) {
      this._calibrateWithDuration(
        this.options.cueDurationMaximum,
        this.options
      );
    }
  };

  _incrementLook = () => {
    if (!this.state.isCalibrating) {
      return;
    }
    // Punish.
    if (++this.state.lookNumber >= this.options.lookCriterion) {
      this._calibrateWithDuration(this.state.cueDuration, this.options);
    }
  };

  _didSaccade() {
    const { didSaccade } = this.delegate;
    isFunction(didSaccade) &&
      didSaccade({ ...this.getState(), isSaccadeAway: this.isSaccadeAway });
  }

  _didSaccadeEarly() {
    this.state.reward = false;
    if (typeof this.isSaccadeAway !== 'undefined') {
      this._didSaccade();
    }
  }

  _nextTrial = () => {
    if (!this.state.isStarted) {
      return;
    }
    clearInterval(this.animationTimer);

    this.isSaccadeAway = undefined;
    this.state.reward = true;
    this.keyDelegate = ::this._didSaccadeEarly;

    // After 1-2s random delay, pause the animation and show the cue.
    setTimeout(() => {
      const {
        didAnimationStart,
        didAnimationStop,
        didCueChange,
        didStateChange,
        didSaccade,
        didNotSaccade
      } = this.delegate;

      // Freeze animation and show cue.
      this.state.isAnimating = false;
      this.state.isDistracting = true;
      this.isSaccadeAway = false;
      this.cueTime = window.performance.now();
      isFunction(didCueChange) && didCueChange(this.getState());
      isFunction(didAnimationStop) && didAnimationStop(this.getState());
      isFunction(didStateChange) && didStateChange(this.getState());

      // After 1000ms, don't report saccades as looks to cue.
      setTimeout(
        () => (this.isSaccadeAway = true),
        this.getSaccadeAwayThreshold()
      );

      // After the cue duration expires, hide the distractor.
      setTimeout(() => {
        this.state.isDistracting = false;
        isFunction(didCueChange) &&
          didCueChange({
            ...this.getState(),
            cueTimeElapsed: Math.floor(window.performance.now() - this.cueTime)
          });

        // After 600ms, we know what kind of saccade the subject made.
        setTimeout(() => {
          this.keyDelegate = ::this._didSaccade;
          this.state.isAnimating = this.state.reward;
          if (this.state.reward) {
            isFunction(didNotSaccade) && didNotSaccade(this.getState());
            this._incrementCueDuration();
          } else {
            this._incrementLook();
          }
          isFunction(didStateChange) && didStateChange(this.getState());
        }, this.getSaccadeThreshold());

        // After all work is complete, reset and restart the animation.
        setTimeout(() => {
          if (
            this.state.trialNumberAtCalibrationEnd &&
            this.state.trialNumber >= this.state.trialNumberAtCalibrationEnd
          ) {
            this.stop();
            return;
          }

          if (this.hasFocus) {
            this._nextTrial();
          } else {
            this.nextTrial = once(::this._nextTrial);
          }
          clearInterval(this.animationTimer);
          this.state.isAnimating = true;
          this.state.trialNumber++;

          this.state.cueDirection = this.nextDirectionAsBoolean()
            ? 'left'
            : 'right';

          this.isStarted() &&
            isFunction(didAnimationStart) &&
            didAnimationStart(this.getState());

          isFunction(didStateChange) && didStateChange(this.getState());
          if (!this.hasFocus && this.state.trialNumber % 2 === 0) {
            this.animationTimer = this.createAnimationTimer();
          }
        }, 3000);
      }, this.state.cueDuration);
    }, this.getDelayInMilliseconds());
  };
}
