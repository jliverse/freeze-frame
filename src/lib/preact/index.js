import partialRight from 'lodash/partialRight';
import { h, render } from 'preact';
import { render as renderAsHTML } from 'preact-render-to-string';

/**
 * Mount a component and return an unmount function for testing.
 */
export function mount(component) {
  return new Promise((resolve, reject) => {});
  const root = render(
    component,
    document.body,
    document.body.firstElementChild
  );
  return () => {
    render('', document.body, root);
  };
}

/**
 * Mount a component for testing.
 */
export function mountAsPromise(component, fn) {
  return new Promise((resolve, reject) => {
    const root = render(
      component,
      document.body,
      document.body.firstElementChild
    );
    defer(() => {
      resolve(root);
    });
  });
}

export function isolateWithFunction(fn) {
  return e => {
    if (!e.target.checked) {
      e.preventDefault();
    }
    e.stopPropagation();
    fn(e);
    return false;
  };
}

const deferWithPromise = Promise.prototype.then.bind(Promise.resolve());

const defer = partialRight(setTimeout, 0);

export { defer, deferWithPromise, renderAsHTML };
