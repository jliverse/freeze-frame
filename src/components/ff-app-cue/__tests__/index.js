import { h, render } from 'preact';
import { deferWithPromise as defer, renderAsHTML } from 'lib/preact';

const log = require('debug')('ff:app:cue');

import UI from '../index';

test('UI renders', () => {
  const html = renderAsHTML(<UI />);
  expect(html).toBeTruthy();
  expect(html).toMatchSnapshot();
});

test('UI accepts a direction (left)', () => {
  const ui = <UI direction="left" size="16" />;
  const html = renderAsHTML(ui);
  expect(html).toMatchSnapshot();
});

test('UI accepts a direction (right)', () => {
  const ui = <UI direction="right" size="16" />;
  const html = renderAsHTML(ui);
  expect(html).toMatchSnapshot();
});
