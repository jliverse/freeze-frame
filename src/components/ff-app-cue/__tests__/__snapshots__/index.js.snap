// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI accepts a direction (left) 1`] = `"<aside class=\\"ff-app-cue dn\\" style=\\"transform: translateX(0px); xleft: 0; width: 16px; height: 16px\\"></aside>"`;

exports[`UI accepts a direction (right) 1`] = `"<aside class=\\"ff-app-cue dn\\" style=\\"transform: translateX(0px); xright: 0; width: 16px; height: 16px\\"></aside>"`;

exports[`UI renders 1`] = `"<aside class=\\"ff-app-cue dn\\" style=\\"transform: translateX(0px); xleft: 0; width: 1px; height: 1px\\"></aside>"`;
